### ISSUE TITLE
Windows desktop capture performance

### ISSUE THREAD

**Issue by bp2008** [Ylianst/MeshCentral#2003]
Hi.  First of all, thank you so much for building MeshCentral.  I just discovered it today, and I am in awe of what you have accomplished.  In 2016, I started a much less ambitious [project](https://github.com/bp2008/SelfHostedRemoteDesktop) with remarkably similar goals and design choices, but I never got very far due to a lack of time and motivation.  I am thrilled to see so many of my goals fully realized by MeshCentral, and so much more.  With such active development too! ❤️❤️❤️

So, I set up a MeshCentral server in an Ubuntu 20.04 VM (which went fine following instructions for 18.04) and installed the agent software on two Windows machines.  Something I noticed right away is that the desktop capture performance is (I'm sorry) terrible.  To be specific, I mean the "fast" frame rate is quite slow, it makes the local mouse pointer flicker, and it consumes significant CPU time when nothing is changing on screen.  Based on earlier issues (#1237, #1208, #357) it seems you are open to improving all of this.

The solution is to use [DXGI Desktop Duplication](https://docs.microsoft.com/en-us/windows/win32/api/dxgi1_2/nn-dxgi1_2-idxgioutputduplication) which is an API available in Windows 8 and newer.  This API tells you about regions of the screen that have [moved](https://docs.microsoft.com/en-us/windows/win32/api/dxgi1_2/nf-dxgi1_2-idxgioutputduplication-getframemoverects) and [changed](https://docs.microsoft.com/en-us/windows/win32/api/dxgi1_2/nf-dxgi1_2-idxgioutputduplication-getframedirtyrects) in a very efficient manner.  The frame rate is limited mainly by your encoding, data streaming, and decoding pipeline.  When [I used this several years ago](https://github.com/bp2008/SelfHostedRemoteDesktop/blob/1aa7254d5fc1eaf0b82c84b75ab8a4870e0ef4d0/SelfHostedRemoteDesktop/Streamer/DxgiOutputDuplicator.cs), I was able to achieve 60+ FPS when dragging a small window around on a 4K desktop.  I would send the moved region information straight to the browser where there are HTML5 canvas operations to easily handle moving image data regions.  For the changed regions, I'd encode them with libjpeg-turbo and have the browser decode them and draw them in the appropriate spots on the canvas.  This jpeg approach won't handle full screen video very well, but for most desktop usage this approach performs pretty well.  To do any better I suspect it would be necessary to use a hardware-accelerated video encoder.

Some caveats about DXGI Desktop Duplication that I recall from when I was messing with it:
* Requires Windows 8 and newer
* It failed when no user was logged in, forcing me to fall back to other capture methods.  I'd have my agent program occasionally try to restart the DXGI code so that the full efficiency and frame rate could be restored after a user login.

I'm not sure, maybe I should have posted this in the MeshAgent repo?

**Comment by Ylianst** [comment]
Bryan has been working on that exact improvement. MeshCentral started a long time ago and we used to support WindowsXP. So, yes, it's going to be nice if we can improve video capture at some point.

**Comment by TerAnYu** [comment]
Hello!
Are you planning a change for Windows? At the moment, I am experiencing a performance problem when connecting to a windows agent, very slowly receiving/transmitting cursor movement. Changing quality from 40% to 5% does not improve the situation.

**Comment by krayon007** [comment]
Yes it's still on my roadmap. I actually put it on my OKR to get it done in the next few quarters. Which windows version are you running? I know the current method uses a bit of CPU, but you shouldnt be receiving cursor updates slowly...

**Comment by TerAnYu** [comment]
@krayon007, i use different versions:
Winfows 10 Pro 21H2 (19044.2130)
Windows 11 Pro - 21H2 (22000.1098)

**Comment by bp2008** [comment]
It looks like when you moved the cursor on the left machine, it was moving smoothly and its movements were being reflected smoothly on the right machine.  The only performance problem I see there is the low frame rate in the web client, which unfortunately is still normal for the Windows mesh agent.

Based on what I remember from looking at the source code 2 years ago, the primary bottleneck was probably the desktop capture method being used.  That is why I recommended DXGI duplication above as a different capture method.  DXGI is extremely fast at desktop capture.  DXGI also tells you about the regions of the desktop that moved or changed, so the mesh agent won't need to waste time looking for changes in every video frame anymore (if it is even bothering to do that, I don't remember).

**Comment by TerAnYu** [comment]
Thanks for the answer!
Yes, the pointer moves smoothly on both sides, even at 64 kbps.
The web client settings were set to:
Quality - 40%
Scaling - 100%
Frame rate-Fast

As far as I understand, it is impossible to switch the image capture mode?

**Comment by bp2008** [comment]
Correct.  As far as I know, there only is one mode implemented and it is quite slow.

**Comment by TerAnYu** [comment]
Have a workaround for the current problem or wait for your solution?

**Comment by dlong500** [comment]
@Ylianst It looks like unfortunately @krayon007 has dropped off the radar over the last couple years.  Is this still on the roadmap to improve desktop capture?

**Comment by marclaporte** [comment]
@dlong500 @TerAnYu @bp2008

- https://github.com/Ylianst/MeshCentral/discussions/4795
- "In November 2022, Intel discontinued its investment in, and is no longer contributing to, the Mesh open-source project." Source: https://www.intel.com/content/www/us/en/content-details/773384/intel-discontinues-contribution-to-mesh-open-source-project.html

This feature request, along with many others, is still on the roadmap. We need more users, organizations and developers to step up. We also need expertise in writing grant requests (proposals).

My company @EvoluData and the @WikiSuite project have stepped up and we are working on a UX revamp: https://github.com/Ylianst/MeshCentral/issues/6333

I urge everyone to join me in sponsoring Simon Smith (si458), who is doing great work!
https://github.com/sponsors/si458

"Probably by far the most amazing development when it comes to MeshCentral is the amazing work Simon Smith (si458) how submitted a lot of PR's and is motivating me to start moving forward with more updates. Thanks Simon for the ongoing work." Source: https://meshcentral2.blogspot.com/2023/10/meshcentral-windows-arm64-nodejs-v11.html

If you'd like to sponsor the development of this or any feature, please reach out to @si458 or myself.

**Comment by mdalacu** [comment]
Nothing new on this front?
Thanks
